src/core/server/execution_context/execution_context_service.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import { AsyncLocalStorage } from 'async_hooks';
+
+import type { CoreService, KibanaExecutionContext } from '../../types';
+import type { CoreContext } from '../core_context';
+import type { Logger } from '../logging';
+
+import { ExecutionContextContainer, getParentContextFrom } from './execution_context_container';
+
+/**
+ * @public
+ */
+export interface KibanaServerExecutionContext extends Partial<KibanaExecutionContext> {
+  requestId: string;
+}
+
+/**
+ * @internal
+ */
+export interface IExecutionContext {
+  getParentContextFrom(headers: Record<string, string>): KibanaExecutionContext | undefined;
+  set(context: Partial<KibanaServerExecutionContext>): void;
+  reset(): void;
+  get(): ExecutionContextContainer | undefined;
+}
+
+/**
+ * @internal
+ */
+export type InternalExecutionContextSetup = IExecutionContext;
+
+/**
+ * @internal
+ */
+export type InternalExecutionContextStart = IExecutionContext;
+
+/**
+ * @public
+ */
+export interface ExecutionContextSetup {
+  /**
+   * Stores the meta-data of a runtime operation.
+   * Data are carried over all async operations automatically.
+   * The sequential calls merge provided "context" object shallowly.
+   **/
+  set(context: Partial<KibanaServerExecutionContext>): void;
+  /**
+   * Retrieves an opearation meta-data for the current async context.
+   **/
+  get(): ExecutionContextContainer | undefined;
```

Comment:
Used for testing only. Maybe we can remove it from the public contract (and keep it for internal purposes).